File: ColorSpaces/interactiveColorSegment.cpp
#include "color_spaces.hpp"

#include <opencv2/core.hpp>
#include <opencv2/highgui.hpp>
#include <opencv2/imgproc.hpp>

#include <array>
#include <cctype>
#include <filesystem>
#include <iostream>
#include <optional>
#include <stdexcept>
#include <string>

#ifndef COLOR_SPACES_PROJECT_DIR
#error "COLOR_SPACES_PROJECT_DIR must be defined by CMake"
#endif

namespace {

int parseInteger(const std::string& text) {
    std::size_t consumed = 0;
    const int value = std::stoi(text, &consumed);
    if (consumed != text.size()) {
        throw std::invalid_argument("invalid integer: " + text);
    }
    return value;
}

struct Options {
    std::filesystem::path input =
        std::filesystem::path(COLOR_SPACES_PROJECT_DIR) / "images" / "rub00.jpg";
    color_spaces::ColorSpace colorSpace = color_spaces::ColorSpace::Hsv;
    std::array<int, 3> lower{20, 80, 40};
    std::array<int, 3> upper{45, 255, 255};
    std::optional<std::filesystem::path> outputDirectory;
    bool noDisplay = false;
    bool validate = false;
    bool help = false;
};

std::array<int, 3> parseTriplet(const int argc, char** argv, int& index) {
    if (index + 3 >= argc) {
        throw std::invalid_argument("threshold option requires three integer values");
    }
    return {
        parseInteger(argv[++index]),
        parseInteger(argv[++index]),
        parseInteger(argv[++index]),
    };
}

Options parseArguments(const int argc, char** argv) {
    Options options;
    for (int index = 1; index < argc; ++index) {
        const std::string argument(argv[index]);
        if (argument == "--input") {
            if (++index >= argc) {
                throw std::invalid_argument("--input requires a path");
            }
            options.input = std::filesystem::absolute(argv[index]);
        } else if (argument == "--space") {
            if (++index >= argc) {
                throw std::invalid_argument("--space requires BGR, HSV, YCrCb, or Lab");
            }
            options.colorSpace = color_spaces::parseColorSpace(argv[index]);
        } else if (argument == "--lower") {
            options.lower = parseTriplet(argc, argv, index);
        } else if (argument == "--upper") {
            options.upper = parseTriplet(argc, argv, index);
        } else if (argument == "--output-dir") {
            if (++index >= argc) {
                throw std::invalid_argument("--output-dir requires a path");
            }
            options.outputDirectory = std::filesystem::absolute(argv[index]);
        } else if (argument == "--no-display") {
            options.noDisplay = true;
        } else if (argument == "--validate") {
            options.validate = true;
        } else if (argument == "--help" || argument == "-h") {
            options.help = true;
        } else {
            throw std::invalid_argument("unknown argument: " + argument);
        }
    }
    return options;
}

struct Segmentation {
    cv::Mat mask;
    cv::Mat result;
};

Segmentation segment(const cv::Mat& image, const Options& options) {
    cv::Mat mask = color_spaces::thresholdMask(
        image, options.colorSpace, options.lower, options.upper);
    cv::Mat result = color_spaces::applyMask(image, mask);
    return Segmentation{mask, result};
}

std::pair<std::filesystem::path, std::filesystem::path> outputPaths(
    const Options& options) {
    if (!options.outputDirectory.has_value()) {
        throw std::logic_error("output directory is unavailable");
    }
    const std::string stem = options.input.stem().string();
    std::string normalizedSpace = color_spaces::name(options.colorSpace);
    for (char& character : normalizedSpace) {
        character = static_cast<char>(
            std::tolower(static_cast<unsigned char>(character)));
    }
    return {
        *options.outputDirectory / (stem + "-" + normalizedSpace + "-mask.png"),
        *options.outputDirectory / (stem + "-" + normalizedSpace + "-result.png"),
    };
}

void writeOutputs(const Options& options, const Segmentation& segmentation) {
    const auto [maskPath, resultPath] = outputPaths(options);
    color_spaces::writeImage(maskPath, segmentation.mask);
    color_spaces::writeImage(resultPath, segmentation.result);
    std::cout << "Wrote: " << maskPath << '\n';
    std::cout << "Wrote: " << resultPath << '\n';
}

void runValidation(const Options& options) {
    const cv::Mat image = color_spaces::readBgr(options.input);
    const Segmentation segmentation = segment(image, options);
    const int foregroundPixels = cv::countNonZero(segmentation.mask);
    if (foregroundPixels <= 0 ||
        foregroundPixels >= segmentation.mask.rows * segmentation.mask.cols) {
        throw std::runtime_error("foreground pixel count is outside the expected range");
    }
    if (segmentation.result.size() != image.size() ||
        segmentation.result.type() != image.type()) {
        throw std::runtime_error("segmentation result does not match input shape/type");
    }
    if (options.outputDirectory.has_value()) {
        writeOutputs(options, segmentation);
    }
    std::cout << "VALIDATION PASSED: foreground_pixels=" << foregroundPixels
              << ", image=" << image.cols << 'x' << image.rows << '\n';
}

void printUsage(const char* program) {
    std::cout
        << "Usage: " << program
        << " [--input IMAGE] [--space BGR|HSV|YCrCb|Lab]"
        << " [--lower C0 C1 C2] [--upper C0 C1 C2]"
        << " [--output-dir DIR] [--no-display] [--validate]\n";
}

}  // namespace

int main(const int argc, char** argv) {
    try {
        const Options options = parseArguments(argc, argv);
        if (options.help) {
            printUsage(argv[0]);
            return 0;
        }
        if (options.validate) {
            runValidation(options);
            return 0;
        }

        const cv::Mat image = color_spaces::readBgr(options.input);
        const Segmentation segmentation = segment(image, options);
        const int foregroundPixels = cv::countNonZero(segmentation.mask);
        std::cout << "Input: " << options.input << '\n';
        std::cout << "Color space: " << color_spaces::name(options.colorSpace) << '\n';
        std::cout << "Foreground pixels: " << foregroundPixels << '/'
                  << segmentation.mask.rows * segmentation.mask.cols << '\n';

        if (options.outputDirectory.has_value()) {
            writeOutputs(options, segmentation);
        }
        if (!options.noDisplay) {
            cv::imshow("Input", image);
            cv::imshow("Mask", segmentation.mask);
            cv::imshow(
                color_spaces::name(options.colorSpace) + " segmentation",
                segmentation.result);
            cv::waitKey(0);
            cv::destroyAllWindows();
        }
        return 0;
    } catch (const std::exception& error) {
        std::cerr << "Error: " << error.what() << '\n';
        printUsage(argv[0]);
        return 2;
    }
}

File: ColorSpaces/interactiveColorSegment.py
#!/usr/bin/env python3
"""Segment a BGR image with thresholds defined in a chosen color space."""

from __future__ import annotations

import argparse
from collections.abc import Sequence
from pathlib import Path

import cv2
import numpy as np

from color_spaces import (
    COLOR_CONVERSIONS,
    apply_mask,
    read_bgr,
    resolve_input_path,
    run_core_validation,
    threshold_mask,
    write_image,
)


def segment_image(
    image: np.ndarray,
    color_space: str,
    lower: Sequence[int],
    upper: Sequence[int],
) -> tuple[np.ndarray, np.ndarray]:
    """Return a one-channel mask and a BGR visualization."""

    mask = threshold_mask(image, color_space, lower, upper)
    return mask, apply_mask(image, mask)


def output_paths(output_dir: str | Path, stem: str, color_space: str) -> tuple[Path, Path]:
    """Return deterministic mask and visualization output paths."""

    directory = Path(output_dir).expanduser().resolve()
    normalized_space = color_space.lower()
    return (
        directory / f"{stem}-{normalized_space}-mask.png",
        directory / f"{stem}-{normalized_space}-result.png",
    )


def run_validation(output_dir: str | Path | None = None) -> dict[str, int]:
    """Validate the default yellow segmentation on a bundled image."""

    metrics = run_core_validation()
    input_path = resolve_input_path(None)
    image = read_bgr(input_path)
    mask, result = segment_image(image, "HSV", (20, 80, 40), (45, 255, 255))
    foreground_pixels = int(cv2.countNonZero(mask))
    if foreground_pixels <= 0 or foreground_pixels >= mask.size:
        raise AssertionError(f"unexpected foreground pixel count: {foreground_pixels}")
    if result.shape != image.shape:
        raise AssertionError("segmentation result shape differs from input")

    if output_dir is not None:
        mask_path, result_path = output_paths(output_dir, input_path.stem, "HSV")
        write_image(mask_path, mask)
        write_image(result_path, result)

    print(
        "VALIDATION PASSED: "
        f"foreground_pixels={foreground_pixels}, "
        f"image={image.shape[1]}x{image.shape[0]}"
    )
    return {**metrics, "foreground_pixels": foreground_pixels}


def _triplet(values: list[str]) -> tuple[int, int, int]:
    """Parse three integer threshold values for argparse."""

    try:
        parsed = tuple(int(value) for value in values)
    except ValueError as error:
        raise argparse.ArgumentTypeError("thresholds must be integers") from error
    if len(parsed) != 3:
        raise argparse.ArgumentTypeError("thresholds require three values")
    return parsed  # type: ignore[return-value]


def _build_parser() -> argparse.ArgumentParser:
    """Create the command-line parser."""

    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument("--input", type=Path, help="input image; defaults to images/rub00.jpg")
    parser.add_argument(
        "--space",
        choices=tuple(COLOR_CONVERSIONS),
        default="HSV",
        help="color space in which thresholds are interpreted",
    )
    parser.add_argument(
        "--lower",
        nargs=3,
        default=("20", "80", "40"),
        metavar=("C0", "C1", "C2"),
        help="inclusive lower threshold",
    )
    parser.add_argument(
        "--upper",
        nargs=3,
        default=("45", "255", "255"),
        metavar=("C0", "C1", "C2"),
        help="inclusive upper threshold",
    )
    parser.add_argument("--output-dir", type=Path, help="directory for mask and result PNGs")
    parser.add_argument("--no-display", action="store_true", help="do not open GUI windows")
    parser.add_argument("--validate", action="store_true", help="run deterministic checks")
    return parser


def main(argv: Sequence[str] | None = None) -> int:
    """Run segmentation and save reproducible outputs."""

    arguments = _build_parser().parse_args(argv)
    if arguments.validate:
        run_validation(arguments.output_dir)
        return 0

    lower = _triplet(arguments.lower)
    upper = _triplet(arguments.upper)
    input_path = resolve_input_path(arguments.input)
    image = read_bgr(input_path)
    mask, result = segment_image(image, arguments.space, lower, upper)
    foreground_pixels = int(cv2.countNonZero(mask))
    print(f"Input: {input_path}")
    print(f"Color space: {arguments.space}")
    print(f"Foreground pixels: {foreground_pixels}/{mask.size}")

    if arguments.output_dir is not None:
        mask_path, result_path = output_paths(
            arguments.output_dir, input_path.stem, arguments.space
        )
        write_image(mask_path, mask)
        write_image(result_path, result)
        print(f"Wrote: {mask_path}")
        print(f"Wrote: {result_path}")

    if not arguments.no_display:
        cv2.imshow("Input", image)
        cv2.imshow("Mask", mask)
        cv2.imshow(f"{arguments.space} segmentation", result)
        cv2.waitKey(0)
        cv2.destroyAllWindows()
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
